src/Network.h

```diff
@@ -79,43 +72,78 @@ class Network {
-    static std::pair<int, int> load_network_file(const std::string& filename);
-    static void process_bn_var(std::vector<float>& weights,
+    std::pair<int, int> load_v1_network(std::istream& wtfile);
+    std::pair<int, int> load_network_file(const std::string& filename);
+    void process_bn_var(std::vector<float>& weights,
                                const float epsilon = 1e-5f);
 
-    static std::vector<float> winograd_transform_f(const std::vector<float>& f,
+    std::vector<float> winograd_transform_f(const std::vector<float>& f,
         const int outputs, const int channels);
-    static std::vector<float> zeropad_U(const std::vector<float>& U,
+    std::vector<float> zeropad_U(const std::vector<float>& U,
         const int outputs, const int channels,
         const int outputs_pad, const int channels_pad);
-    static void winograd_transform_in(const std::vector<float>& in,
+    void winograd_transform_in(const std::vector<float>& in,
                                       std::vector<float>& V,
                                       const int C);
-    static void winograd_transform_out(const std::vector<float>& M,
+    void winograd_transform_out(const std::vector<float>& M,
                                        std::vector<float>& Y,
                                        const int K);
-    static void winograd_convolve3(const int outputs,
+    void winograd_convolve3(const int outputs,
                                    const std::vector<float>& input,
                                    const std::vector<float>& U,
                                    std::vector<float>& V,
                                    std::vector<float>& M,
                                    std::vector<float>& output);
-    static void winograd_sgemm(const std::vector<float>& U,
+    void winograd_sgemm(const std::vector<float>& U,
                                const std::vector<float>& V,
                                std::vector<float>& M, const int C, const int K);
+    Netresult get_scored_moves_internal(const GameState* const state,
+                                               const int symmetry);
+
     static void fill_input_plane_pair(const FullBoard& board,
                                       std::vector<net_t>::iterator black,
                                       std::vector<net_t>::iterator white,
                                       const int symmetry);
-    static Netresult get_scored_moves_internal(const GameState* const state,
-                                               const int symmetry);
+
+    bool probe_cache(const GameState* const state, Network::Netresult& result);
 #if defined(USE_BLAS)
-    static void forward_cpu(const std::vector<float>& input,
+    void forward_cpu(const std::vector<float>& input,
                             std::vector<float>& output_pol,
                             std::vector<float>& output_val);
 
 #endif
-};
 
+#ifdef USE_OPENCL
+    OpenCLScheduler m_opencl;
```

Comment:
It feels a bit weird to have this owned by Network. I guess it's really here because it contains pointers to contexts which themselves identify where on the GPUs the network data is loaded.

Reply:
I agree with you, but I couldn't find a better place for this.  I have another batch of fixes (abstract interface for network evaluation backend) and this is replaced with a pointer to the abstract type.  This piece needs a bit more clever thought but no good idea yet...